src/ProgressLogging.jl

```diff
@@ -3,13 +3,168 @@ module ProgressLogging
+  This is used for representing progresses of nested tasks.  Note that
+  sub-tasks may be executed concurrently; i.e., there can be multiple
+  child tasks for one parent task.
+- `name::String`: Name of the progress bar.
+- `done::Bool`: `true` if the task is done.
+"""
+Base.@kwdef struct Progress
+    id::UUID
+    parentid::UUID = ROOTID  # not nested by default
+    fraction::Float64 = NaN
+    name::String = ""
+    done::Bool = false
+end
+
+Progress(id::UUID, fraction::Real = NaN; kwargs...) =
+    Progress(; kwargs..., fraction = fraction, id = id)
+
+# Define `string`/`print` so that progress log records are (somewhat)
+# readable even without specific log monitors.
+function Base.print(io::IO, progress::Progress)
+    print(io, isempty(progress.name) ? "Progress" : progress.name)
+    if progress.parentid !== ROOTID
+        print(io, " (sub)")
+    end
+    print(io, ": ")
+    if isnan(progress.fraction)
+        print(io, "??%")
+    else
+        print(io, floor(Int, progress.fraction * 100), '%')
+    end
+    return
+end
+
+const PROGRESS_LOGGING_UUID_NS = UUID("1e962757-ea70-431a-b9f6-aadf988dcb7f")
+
+asuuid(id::UUID) = id
+asuuid(id) = uuid5(PROGRESS_LOGGING_UUID_NS, string(id))
+
+asfraction(progress::Nothing) = NaN
+asfraction(progress::AbstractString) = progress == "done" ? NaN : nothing
+asfraction(progress::Real) = convert(Float64, progress)
+
+
+"""
+    ProgressLogging.asprogress(_, name, _, _, id, _, _; progress, ...) :: Union{Progress, Nothing}
+
+Pre-process log record to obtain a [`Progress`](@ref) object if it is
+one of the supported format.  This is mean to be used with the
+`message` positional argument and _all_ keyword arguments passed to
+`Logging.handle_message`.  Example:
+
+```julia
+function Logging.handle_message(logger::MyLogger, args...; kwargs...)
+    progress = ProgressLogging.asprogress(args...; kwargs...)
+    if progress !== nothing
+        return # handle progress log record
+    end
+    # handle normal log record
+end
+```
+"""
+asprogress(_level, progress::Progress, _args...; _...) = progress
+asprogress(_level, name, _module, _group, id, _file, _line; progress = undef, kwargs...) =
+    if progress isa Union{Nothing,Real,AbstractString}
+        _asprogress(name, id; progress = progress, kwargs...)
+    else
+        nothing
+    end
+
+# `parentid` is used from `@logprogress`.
+function _asprogress(name, id, parentid = ROOTID; progress, _...)
+    fraction = asfraction(progress)
+    fraction === nothing && return nothing
+    return Progress(
+        fraction = fraction,
+        name = name,
+        id = asuuid(id),
+        parentid = parentid,
+        done = progress == "done"
+    )
+end
+
+# To pass `Progress` value without breaking progress monitors with the
+# previous `progress` key based specification, we create an abstract
+# string type that has `Progress` attached to it.  This is used as the
+# third argument `message` of `Logging.handle_message`.
```

Comment:
Clever!  Would the plan be to remove this and just use bare `Progress` in the future?

Reply:
I think @pfitzseb wants to keep the original open/untyped API.  I'm OK for keeping this as long as this compatible layer becomes annoying hard to maintain.  `ProgressString` is somewhat already "too clever" but maybe it's within an OK range.